test: add and use prefix substitutions for libraries

compnerd opened the pull request:
The naming convention is different on Windows than on Unix-like
environments.  In order to follow the convention we need to substitute
the prefix and the suffix.  Take the opportunity to rename the
`target-dylib-extension` to the CMake-like variable
`target-shared-library-suffix` and introduce
`target-shared-library-prefix`.  This helps linking the test suite
binaries on Windows.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd committed:
test: add and use prefix substitutions for libraries

The naming convention is different on Windows than on Unix-like
environments.  In order to follow the convention we need to substitute
the prefix and the suffix.  Take the opportunity to rename the
`target-dylib-extension` to the CMake-like variable
`target-shared-library-suffix` and introduce
`target-shared-library-prefix`.  This helps linking the test suite
binaries on Windows.

compnerd commented:
@swift-ci please test

jrose-apple commented on test/Interpreter/SDK/objc_getClass.swift:
```diff
@@ -1,15 +1,15 @@
 
-// RUN: %target-build-swift-dylib(%t/libresilient_struct.%target-dylib-extension) -Xfrontend -enable-resilience -Xfrontend -enable-class-resilience %S/../../Inputs/resilient_struct.swift -emit-module -emit-module-path %t/resilient_struct.swiftmodule -module-name resilient_struct
-// RUN: %target-codesign %t/libresilient_struct.%target-dylib-extension
+// RUN: %target-build-swift-dylib(%t/%{target-shared-library-prefix}resilient_struct%{target-shared-library-suffix}) -Xfrontend -enable-resilience -Xfrontend -enable-class-resilience %S/../../Inputs/resilient_struct.swift -emit-module -emit-module-path %t/resilient_struct.swiftmodule -module-name resilient_struct
```
This pattern is very verbose. Maybe we should make it another regex-based substitution? `%target-shared-library(resilient_struct)`.

compnerd replied:
I tried to do that, but failed at that, but I thought it would be possible to do that substitution.  But, since you also seem to think that it should be possible to have that substitution, I want to give that a second shot.

compnerd committed:
test: create and use target-library-name

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@jrose-apple - so, I ran into the same problem as last time, the substitution within the capture doesn't work right and I can't seem to identify how to get that to work (even with a braced version, it doesn't seem to do the substitution correctly).  However, wherever possible, I am now using the `%target-library-name` substitution with a capture instead of spelling it out.

compnerd commented:
@swift-ci please test

jrose-apple commented:
lit substitutions are ordered, so just to check: did you try putting your new substitution either before or after the existing substitution, whichever one you hadn't done yet?

compnerd commented:
That does the trick!  Thanks!

compnerd committed:
test: make `%target-library-name` work in captures

Thanks to @jrose for the hint about the substitution ordering, the new
substitution now works even inside the capture group.  Replace the
remaining uses to the new macro.

compnerd commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test and merge
